Ubuntu 16.04 compatibility fix follow up

chris-allan opened the pull request:
This is a follow up to #3455 which fixes a runtime dependency resolution issue that manifests itself when having the `decorator` dependency explicitly set (2910fe126d75f00f0de14b2377807ae211093678) and following the [Ubuntu 16.04](https://github.com/CellProfiler/CellProfiler/wiki/Source-installation-%28Ubuntu-16.04-LTS%29) install documentation.

Everything will install fine, but you will get this exception when trying to run CellProfiler:

```
Traceback (most recent call last):
  File "/home/callan/v/cp22-new/bin/cellprofiler", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2927, in <module>
    @_call_aside
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2913, in _call_aside
    f(*args, **kwargs)
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2940, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 637, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 650, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/callan/v/cp22-new/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 834, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (decorator 4.0.6 (/usr/lib/python2.7/dist-packages), Requirement.parse('decorator>=4.1.0'), set(['networkx']))
```

When following the aforementioned documentation, decorator 4.0.6 will have been installed globally by APT as a dependency of scipy. For some reason that I'm not 100% on it will not be upgraded even though networkx 2.0 depends on decorator >= 4.1.0. networkx 2.0 will be installed by `pip` when `pip install --editable .` is executed following this dependency tree:

 * `networkx>=1.8 (from scikit-image->centrosome>=1.0.4->CellProfiler==2.2.0)`

The original change (2910fe126d75f00f0de14b2377807ae211093678) was made to fix Travis running Ubuntu 14.04 with networkx installed globally via APT. This version is 1.8.1 which includes no dependency on the decorator package. As no explicit dependency was then included in CellProfiler's `setup.py` imports of the `cellprofiler.objects` package would fail.

It suffices to simply remove the global APT installation of networkx. networkx will then be installed via `pip` following the aforementioned dependency tree along with the decorator dependency.

chris-allan commented:
I note that a similar situation is present on Ubuntu 14.04 between six, pytest, and h5py and can manifest itself as the following exception when trying to run CellProfiler:

```
Traceback (most recent call last):
  File "/home/cp/venv/bin/cellprofiler", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3147, in <module>
    @_call_aside
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3131, in _call_aside
    f(*args, **kwargs)
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3160, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 668, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 681, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 870, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'six>=1.10.0' distribution was not found and is required by pytest
```

six 1.5.1 is installed globally by APT as a dependency of several packages on Ubuntu 14.04. It is a versionless dependency of h5py and a versioned >= 1.10.0 dependency of pytest. The only way I have been able to "fix" this resolution consistently is to change the order of the dependencies so that pytest is resolved before h5py in `setup.py`:

```
diff --git a/setup.py b/setup.py
index 8abfd77..7731542 100644
--- a/setup.py
+++ b/setup.py
@@ -428,6 +428,7 @@ setuptools.setup(
         "cellh5",
         "centrosome>=1.0.4",
         "decorator",
+        "pytest",
         "h5py<2.7.0",
         "inflect",
         "javabridge",
@@ -436,7 +437,6 @@ setuptools.setup(
         "MySQL-python",
         "numpy<1.12.0",
         "prokaryote",
-        "pytest",
         "python-bioformats",
         "pyzmq",
         "scipy"
```

This feels supremely dirty and brittle.

Given that the [Ubuntu 14.04](https://github.com/CellProfiler/CellProfiler/wiki/Source-Installation-%28Ubuntu-14.04-LTS%29) install documentation explicitly states to upgrade six I am inclined to leave this particular dependency resolution idiosyncrasy as an exercise to the reader.

The reason this problem is not seen on Travis is that the virtualenv that is provided as part of the build execution environment already has six upgraded. As of this writing the version installed is 1.11.0.

mcquin commented:
@chris-allan pytest should only be required for testing. Could you drop that dependency from the `install_requires` block?

AetherUnbound commented:
If it's not there already, it could go in the `tests_require` block instead.

chris-allan commented:
I was very reluctant to modify `setup.py` more significantly but I guess I'll just have to bite the bullet. The reference in `setup_requires` is required due to the way that pytest is called. I don't think going down the rabbit hole of refactoring that is warranted at this point. At @AetherUnbound's suggestion I added it to `tests_require` for good measure. It's insufficient but can't hurt.

For what its worth, on Ubuntu 14.04 this has now moved the weird six dependency interplay to matplotlib instead:

```
Traceback (most recent call last):
  File "/home/cp/venv/bin/cellprofiler", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3147, in <module>
    @_call_aside
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3131, in _call_aside
    f(*args, **kwargs)
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3160, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 668, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 681, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/cp/venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 870, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'six>=1.10' distribution was not found and is required by matplotlib
```

Again, all the documentation says to update six anyway so I don't think reordering things in `install_requires` is really warranted.
